# 01. Importing Libraries

In [2]:
import pandas as pd
import numpy  as np
import os

# 02. Importing Data

In [3]:
# Create Path
path = r'C:\Users\esteb\OneDrive\Desktop\2025_Instacart Basket Analysis'

In [4]:
# Order Data
df_ords = pd.read_csv(os.path.join(path, '02_Data', 'Original Data', 'orders.csv' ))

In [5]:
# Products Data
df_prods = pd.read_csv(os.path.join(path, '02_Data', 'Original Data', 'products.csv' ))

In [6]:
# Departments Data
df_deps = pd.read_csv(os.path.join(path, '02_Data', 'Original Data', 'departments.csv' ))

# 03. Data Wrangling

In [7]:
df_prods.shape

(49693, 5)

In [7]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [8]:
# Eliminate eval_set column
df_ords = df_ords.drop(columns = ['eval_set'])

In [9]:
# Changing Data Type for Values to exclude
df_ords['order_id'] = df_ords['order_id'].astype('str')
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [10]:
# Changing Column Name
df_ords.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [11]:
df_ords.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329       1             1                  2                  8   
1  2398795       1             2                  3                  7   
2   473747       1             3                  3                 12   
3  2254736       1             4                  4                  7   
4   431534       1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [12]:
# Frequency of Busiest Hour to place order
df_ords['order_hour_of_day'].value_counts()

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

The busiest hour of the day to place orders is at 10am.

# 04. Transposing Data

In [13]:
# Taking a look at columns
df_deps.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [14]:
# Execute (Transposing df_dep)
df_deps_t = df_deps.T

In [15]:
# Take the first row of df_deps_t for the header
new_header = df_deps_t.iloc[0]

In [16]:
new_header

0    department
Name: department_id, dtype: object

In [17]:
# New Dataframe after Row 1
df_deps_t_new = df_deps_t[1:]

In [18]:
# Set the New Header to the Dataframe
df_deps_t_new.columns = new_header

In [19]:
df_deps_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# 05. Data Dictionary

In [20]:
# Assign a value to data_dict
data_dict = df_deps_t_new.to_dict('index')

In [21]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

The department_id 4 is equal to produce department.

# 06. Subset

In [22]:
# Subset to only view Data from the Breakfast Department
df_breakfast =  df_prods[df_prods['department_id']==14]

In [23]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [24]:
# Subset of Dinner Parties using various department_id
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5, 7, 12, 20])]

In [25]:
# New Dataframe of department_id (5, 7, 12, 20)
df_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

By executing the new dataframe name df_dinner_parties it will show me a preview and a total amount of rows and columns in that new dataframe. The new dataframe has 7650 Rows.

In [26]:
# Alternate way of showing the amount of rows and columns without displaying a preview.
df_dinner_parties.shape

(7650, 5)

In [27]:
# Subset of Customer using user_id 1
df_customer_1 = df_ords[df_ords['user_id']== '1']

In [28]:
# Stats about user_id 1
df_customer_1.describe()

order_number  order_day_of_week  order_hour_of_day  \
count     11.000000          11.000000          11.000000   
mean       6.000000           2.636364          10.090909   
std        3.316625           1.286291           3.477198   
min        1.000000           1.000000           7.000000   
25%        3.500000           1.500000           7.500000   
50%        6.000000           3.000000           8.000000   
75%        8.500000           4.000000          13.000000   
max       11.000000           4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

# 07. Exporting Data

In [30]:
# Export Orders Dataframe
df_ords.to_csv(os.path.join(path, '02_Data', 'Prepared Data', 'orders_wrangled_task.csv'))

In [56]:
# Export Departments Dataframe
df_deps_t_new.to_csv(os.path.join(path, '02_Data', 'Prepared Data', 'departments_wrangled_task.csv'))